lib/data_update_scripts/20211115154021_nullify_invalid_tag_fields.rb

```diff
@@ -0,0 +1,34 @@
+module DataUpdateScripts
+  class NullifyInvalidTagFields
+    def run
+      Tag.where(bg_color_hex: "").or(Tag.where(text_color_hex: "")).find_each do |tag|
+        # presence here selects non-empty strings if set  (avoiding unsetting valid values)
+        # but nullifies empty tag colors
+        tag.update_columns(
+          bg_color_hex: tag.bg_color_hex.presence,
+          text_color_hex: tag.text_color_hex.presence,
+        )
+      end
+
+      # normalize empty string to nil (alias_for? behaved correctly already), this is just cleanup
+      Tag.where(alias_for: "").update(alias_for: nil)
+
+      Tag.where.not(alias_for: nil).find_each do |aliased_tag|
+        # this is expected to not do anything - based on this query in blazer giving 0 rows
+        # (there are 64 with aliases, all valid)
```

Comment:
This may be true for DEV, but we have no guarantees for any other Forem instances, so it's a good thing to check. But again, would it make sense to do it in SQL?

And should we add a constraint to avoid it happening in the future?

Reply:
there is a constraint that prevents you _adding_ an alias_for pointing to a missing tag, and it seems possible you could at some point remove a tag (I don't think we do that) and cause a dangling alias - I couldn't find any in the data I have access to, but I'm sure it's _possible_ and there's no real constraint on `alias_for` in the db.

There's not a validation to prevent _creating_ this situation (alias to a valid tag, remove the pointed to tag, leaving the original tag invalid), but from what I can tell it's not happening (aliases seem to be the preferred fix to tag duplication, rather than deletion, are an admin action so unlikely to be done in bad faith).

I don't see a delete action in the admin tags controller at all - I'm not sure there's a mechanism for someone outside of a console or script to remove a tag (though you _can_ rename one, which _would_ cause this situation).

Reply:
> there is a constraint that prevents you _adding_ an alias_for pointing to a missing tag, and it seems possible you could at some point remove a tag (I don't think we do that) and cause a dangling alias

@djuber Are you talking about an ActiveRecord validation? Because they work like you described but a SQL constraint _does_ prevent removing the target row or modifying the column value.

```
jamie=# create table tags (
  id serial primary key,
  name text unique not null,
  alias_for text references tags(name)
);
CREATE TABLE
jamie=# insert into tags (name, alias_for) values ('target', null), ('alias', 'target');
INSERT 0 2
jamie=# delete from tags where name = 'target';
ERROR:  update or delete on table "tags" violates foreign key constraint "tags_alias_for_fkey" on table "tags"
DETAIL:  Key (name)=(target) is still referenced from table "tags".
jamie=# update tags set name = 'another name' where name = 'target';
ERROR:  update or delete on table "tags" violates foreign key constraint "tags_alias_for_fkey" on table "tags"
DETAIL:  Key (name)=(target) is still referenced from table "tags".
```

> I suspect we _don't_ want to permit name as an updatable tag attribute in the controller if there's no form input to set this - #15382 seeks to add that.

Agreed. And if we do add a constraint for it but want to support renaming a tag we'd want to allow for deferring the constraint check until the end of the transaction so because it involves modifying two different columns of N+1 different rows.

Reply:
hmm... I'm not seeing that fk constraint in development - is it possible that was added in a migration that's no longer available?

```
Forem_development=# \d+ tags
... columns ...

Indexes:
    "tags_pkey" PRIMARY KEY, btree (id)
    "index_tags_on_name" UNIQUE, btree (name)
    "index_tags_on_social_preview_template" btree (social_preview_template)
    "index_tags_on_supported" btree (supported)
Foreign-key constraints:
    "fk_rails_a9dc141dc9" FOREIGN KEY (mod_chat_channel_id) REFERENCES chat_channels(id) ON DELETE SET NULL
    "fk_rails_d11c10a859" FOREIGN KEY (badge_id) REFERENCES badges(id) ON DELETE SET NULL
```